### PR TITLE
pios: add pios_dio digital IO system

### PR DESCRIPTION
**Pull request by mlyle**
For review / to show the concept.  This PR maybe shouldn't merge, until it's actually doing something useful.

Todo:

* [x] Verify config registers on F0
* [x] Verify config registers on F3
* [x] Verify config registers on F4
* [x] Verify drives on F0
* [x] Verify drives on F3
* [x] Verify drives on F4
* [x] Support for configuring alternate functions
* [x] Support for configuring analog
* [x] Support for F1

### CODE REVIEW

File: flight/PiOS/STM32/inc/pios_dio.h [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+
+/**
+ * @brief Configures a GPIO as alternate function.
+ * Note that to configure drive strength, pullups, etc, DIO_output or
+ * DIO_input should be called first.
+ *
+ * @param[in] t The GPIO specifier tag (created with e.g. GPIOA_DIO(3) )
+ * @param[in] alt_func The alternate function specifier.
+ */
+static inline void DIO_altfunc(dio_tag_t t, int alt_func);
```
Comment (tracernz) [inline flight/PiOS/STM32/inc/pios_dio.h]: Why uppercase? I think a verb would be a good idea for all these functions. e.g. `dio_set_altfunc`

File: flight/PiOS/STM32/inc/pios_dio.h [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+ * @param[in] t The GPIO specifier tag
+ */
+static inline void DIO_low(dio_tag_t t);
+
+/**
+ * @brief Sets an output GPIO to a chosen logical level.
+ * @param[in] t The GPIO specifier tag
+ * @param[in] high Whether the GPIO should be high.
+ */
+static inline void DIO_set(dio_tag_t t, bool high);
```
Comment (tracernz) [inline flight/PiOS/STM32/inc/pios_dio.h]: Should match the read call, e.g. `dio_write` and `dio_read`.

**Comment by tracernz** [comment]
It's a shame you can't operate on multiple pins.

File: flight/PiOS/STM32/inc/pios_dio.h (hunk 1 above)
Comment (mlyle) [reply flight/PiOS/STM32/inc/pios_dio.h]: need to decide what we're doing with prefixen in general.  PIOS etc are uppercase prefixes still.. I don't know.

**Comment by mlyle** [comment]
Might be possible to extend / wrap this for multiple in the same IO bank.   99% of what we do can work this way though.

Pushed based on your review feedback / names.

File: flight/PiOS/STM32/inc/pios_dio.h (hunk 1 above)
Comment (glowtape) [reply flight/PiOS/STM32/inc/pios_dio.h]: I can do without the PIOS_ prefix. Plenty of other frameworks and APIs don't qualify their functionality with vendor/framework specific prefixes either.

As far as the casing goes, "orthographically correct" mixed casing sure helps readability, but looks odd when doing shorthands. On the other hand, I prefer shorthands. _dio_altfunc_ is certainly preferable to _DIO_AlternateFunction_, at least when typing out things.

**Comment by mlyle** [comment]
If this is acceptable, will close PR and then cherry-pick into 4wif branch.

**Comment by mlyle** [comment]
Incorporated #1714

**Comment by mlyle** [comment]
Correction;; incorporated to #1682.

**Pull request closed without merging**